packages/workbox-webpack-plugin/src/index.js

```diff
@@ -0,0 +1,181 @@
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+const path = require('path');
+const getEntries = require('./lib/get-manifest-entries-with-webpack');
+const generateManifest = require('./lib/generate-manifest-with-webpack');
+const generateOrCopySW = require('./lib/generate-or-copy-sw');
+const webpackAsset = require('./lib/utils/webpack-asset');
+const copyWorkboxSW = require('./lib/utils/copy-workbox-sw');
+const {setReadFile} = require('./lib/utils/read-file');
+
+/**
+ * Use the instance of this in the plugins array of the webpack config.
+ *
+ * @example <caption>Zero-configuration</caption>
+ * const WorkboxWebpackPlugin = require('workbox-webpack-plugin');
+ * .
+ * module.exports = {
+ *   ...
+ *   plugins: [
+ *    new WorkboxWebpackPlugin(); // WorkboxWebpackPlugin zero-configuration
+ *   ]
+ * };
+ *
+ * @example <caption>Supported webpack options</caption>
+ * const WorkboxWebpackPlugin = require('workbox-webpack-plugin');
+ * .
+ * module.exports = {
+ *   ...
+ *   plugins: [
+ *    new WorkboxWebpackPlugin({
+ *      chunks: ['main'],
+ *      excludeChunks: ['mutable'],
+ *      manifestFilename: 'file-manifest.js',
+ *      swSrc: './src/sw.js', // path to an existing service worker
+ *      filename: 'service-worker.js', // destination filename
+ *     });
+ *   ]
+ * };
+ *
+ * @module workbox-webpack-plugin
+ */
+class WorkboxWebpackPlugin {
+  /**
+   * Creates an instance of WorkboxWebpackPlugin.
+   *
+   * @param {module:workbox-build.Configuration} [config] All the options as
+   *        passed to {@link module:workbox-build.generateSWString}.
+   * @param {Array<String>} [config.chunks] Array of chunk names to use for
+   *        generating the asset manifest. All assets belonging to the provided
+   *        chunk names will be included in the asset manifest. Any chunks that
+   *        are not listed or do not have a name will be removed.
+   * @param {Array<String>} [config.excludeChunks] Array of chunk names to
+   *        exclude from the asset manifest. Any asset beloning to the provided
+   *        chunk names will not be included in the asset manifest. This does
+   *        not affect chunks with no chunk name.
+   * @param {string} [config.filename = 'sw.js'] Name of the service worker file
+   * @param {string} [config.manifestFilename = 'file-manifest[hash].js'] Name
+   *        of the manifest file that will be written to the build directory
+   * @param {string} [config.swSrc] Path to an existing service worker file.
+   *        Will be added to the webpack compilation and prepended with
+   *        importScripts('workbox-sw.js', 'file-manifest.js')
+   */
+  constructor(config = {}) {
+    this.config = config;
+  }
+
+  /**
+   * @private
+   * @return {Object} All workbox configuration options that can be accepted
+   * by {@link module:workbox-build.generateSWString}
+   */
+  get generateSWStringOptions() {
```

Comment:
@jeffposnick this is where we should specify which generateSWString properties should be supported.